panther_analysis_tool/lookup_table.py

```diff
@@ -0,0 +1,283 @@
+        request_payload = json.dumps(
+            {"createLookup":
+             {"name": self.name, "enabled": self.spec["enabled"],
+              "description": self.spec.get("description", ""),
+              "reference": self.spec.get("reference", ""),
+              "lookupSchema": self.spec["lookupSchema"], "logTypeMap": self.spec["logTypeMap"]}}
+        )
+        logging.debug(request_payload)
+
+        response = self.client.invoke(
+            FunctionName=LOOKUP_LAMBDA_NAME,
+            InvocationType="RequestResponse",
+            Payload=request_payload,
+        )
+        response_str = response["Payload"].read().decode("utf-8")
+        response_payload = json.loads(response_str)
+        logging.debug(response_payload)
+
+        status_code = response["StatusCode"]
+        if status_code != 200:
+            logging.warning(
+                "Failed to search for the %s Lookup Table\n\tstatus code: %s\n\terror message: %s",
+                self.name,
+                status_code,
+                response_payload.get("errorMessage", response_payload.get("body")),
+            )
+            return 1
+        return 0
+
+
+    def update(self) -> int:
+        request_payload = json.dumps(
+            {"updateLookup":
+             {"name": self.name, "enabled": self.spec["enabled"],
+              "description": self.spec.get("description", ""),
+              "id": self.lookup_id, "reference": self.spec.get("reference", ""),
+              "lookupSchema": self.spec["lookupSchema"], "logTypeMap": self.spec["logTypeMap"]}}
+        )
+        logging.debug(request_payload)
+
+        response = self.client.invoke(
+            FunctionName=LOOKUP_LAMBDA_NAME,
+            InvocationType="RequestResponse",
+            Payload=request_payload,
+        )
+        response_str = response["Payload"].read().decode("utf-8")
+        response_payload = json.loads(response_str)
+        logging.debug(response_payload)
+        error_msg = response_payload.get("errorMessage", None)
+
+        status_code = response["StatusCode"]
+        if status_code != 200 or error_msg:
+            logging.warning(
+                "Failed to search for the %s Lookup Table\n\tstatus code: %s\n\terror message: %s",
+                self.name,
+                status_code,
+                response_payload.get("errorMessage", response_payload.get("body")),
+            )
+            return 1
+
+        if self.local_file_matches():
+            logging.info("The local data file matches S3 and does not need to be uploaded")
+        else:
+            logging.info("The local data file does not match S3 and needs to be uploaded.")
+            if self.upload_data():
+                return 1
+            if self.upload_from_s3():
+                return 1
+        return 0
+
+
+    def get_input_bucket(self) -> str:
+        logging.info("Fetching the input bucket name")
+        response = self.client.get_function(
```

Comment:
I'm not sure if there is need for granting additional permissions to get the function configuration from another Lambda. Alternatively you can create an endpoint in the lookup-tables-api Lambda that just returns the information that `local_file_matches` needs. In general, there is benefit in keeping the client-side simpler (you don't depend so much on version updates for example), so letting the backend handle more complexity can give you more flexibility in the future.

Reply:
@georgepsarakis yes, good point. However, I could also add the bucket name as an environment variable to the lookup table lambda which is what I was going to do at first but I felt like that would be slightly misleading as the Lambda itself does not need permissions to this bucket. In fact, one of the goals of this PR is to provide an alternate mechanism for uploading files to a lookup table so that we can work around Lambda's payload limit.

But 'saying this out loud', I feel like it's not so bad and is lighter weight to just give the lookup lambda the input bucket name in its configuration (env variables).

Reply:
Actually, just call this endpoint (https://github.com/panther-labs/panther-enterprise/blob/master/api/lambda/logtypes/models_gen.go#L30)  on the logTypes api (PAT should already have permission) that returns an "upload" s3 path. @shaun-sweet is enhancing the LUTs api to be able to handle these.

Reply:
OK, to clarify you're recommending that I use this new endpoint to pull the name of which bucket? The one used by the log processor or a different one? Also, this would require that I wait until @shaun-sweet has merged his changes to master.

Reply:
@rleighton

Reply:
OK, this code will be going away once I package-ize the s3 signing code.